docs/container-register-disks.md

```diff
@@ -0,0 +1,247 @@
+storage. *Pending introduction and implementation of VMGroup object*
+
+## Ephemeral Non Bootable local disk storage
+
+Users want to attach a non-bootable disk to a VM preloaded with data. This data
+could consist of anything. It could be configuration data, rpms, or anything
+else a user might want to distribute with their VMs.
+
+## Future Use Cases
+For the sake of simplicity, this proposal is focusing on the details involved
+with use cases where Immutable VMs are launched using ephemeral storage that is
+not persistent between VM restarts.
+
+Container Registry backed disks can be extended work with stateful block
+storage. This persistent block storage functionality will be built upon the
+foundation laid by this initial proposal.
+
+High level notes related to how container registry disks integrate with
+persistent block storage is provided at the end of this document.
+
+## Implementation 
+
+### Concepts
+
+When a container starts, the container's image is pulled from a remote
+container registry. Each container image consists of multiple layers and only
+the layers that are not already present locally are retrieved from the remote
+registry. 
+
+When a container is being initialized, a union mount filesystem is used to
+combine all the container image's layers together to form the container's
+directory tree. When processes in a container modify contents on the filesystem,
+the processes are not modifying the layers provided by the container image, they
+are modifying a layer of the union filesystem that is tracking the delta between
+the active container and the underlying layers provided by the container image.
+
+The delta between the container image and an active container is ephemeral and
+is automatically freed when container is destroyed. Depending on the union
+mount filesystem implementation used, the delta may be tracked at the block
+level or the file level. When tracked at the file level, the first time a file
+is written to causes the entire file to be copied into the delta layer.
+
+These are the concepts we're building upon to provide an ephemeral disk to a VM
+backed backed by a container image.
+
+### High Level Design
+
+**Standardised KubeVirt VM disk Wrapper**
+
+KubeVirt provides a standardized base wrapper container image that serves up a
+user provided VM disk as a network disk consumable by Libvirt. This base
+container image does not have a VM disk inside of it, it merely contains
+everything necessary for serving up a VM disk in a way Libvirt can consume.
+
+**User Workflow**
+
+Users push VM Images into the container registry using the KubeVirt base
+wrapper container image. By default, the base wrapper container KubeVirt
+provides will serve up any VM disk placed in the /disk directory as a block
+device consumable by libvirt.
+
+Example:
+```
+cat << END > Dockerfile 
+FROM kubevirt.io:disk
+ADD fedora25.qcow2 /disk
+END
+docker build -t vmdisks/fedora25:latest .
+docker push vmdisks/fedora25:latest
```

Comment:
I wonder, if we could somehow deliver the image without a base docker image. By that, I mean still this kind of wrapping, but basically without the `FROM` directory.

Then it is absolutely trivial to create this kind of image and upload it with a virtctl/kubectl call, without even having the need of something like docker on the user machine.

The command could look like this:

```
kubectl upload fedora25.qcow2 -t fedora25:latest
```
It would create the docker standard conform tar archive which represents the only layer of the image and contains as it's only content the qcow2 file. Then we can post it to the registry, and are done. Further, one could still use docker, to do exactly the described steps in the proposal.